src/interpreter/stepper.rs

```diff
@@ -163,37 +179,64 @@ impl<'fncx, 'a, 'b: 'a + 'mir, 'mir, 'tcx: 'b> Stepper<'fncx, 'a, 'b, 'mir, 'tcx
         let block_data = self.mir.basic_block_data(self.block);
         block_data.terminator()
     }
+}
 
-    fn extract_constants(&mut self) {
-        let mut extractor = ConstantExtractor {
-            constants: Vec::new(),
-        };
-        extractor.visit_mir(&self.mir);
-        self.constants.push(extractor.constants);
-    }
+struct ConstantExtractor<'a: 'c, 'b: 'a + 'mir + 'c, 'c, 'mir: 'c, 'tcx: 'a + 'b + 'c> {
+    constants: &'c mut Vec<(ConstantId<'tcx>, Span, Pointer, CachedMir<'mir, 'tcx>)>,
+    span: Span,
+    mir: &'c mir::Mir<'tcx>,
+    fncx: &'c mut FnEvalContext<'a, 'b, 'mir, 'tcx>,
 }
 
-struct ConstantExtractor {
-    constants: Vec<(ConstantId, Span)>,
+impl<'a, 'b, 'c, 'mir, 'tcx> ConstantExtractor<'a, 'b, 'c, 'mir, 'tcx> {
+    fn constant(&mut self, def_id: DefId, substs: &'tcx subst::Substs<'tcx>, span: Span) {
+        if self.fncx.gecx.statics.contains_key(&def_id) {
+            return;
+        }
+        let cid = ConstantId::Static {
+            def_id: def_id,
+            substs: substs,
+        };
+        let mir = self.fncx.load_mir(def_id);
+        let ptr = self.fncx.alloc_ret_ptr(mir.return_ty).expect("there's no such thing as an unreachable static");
+        self.constants.push((cid, span, ptr, mir));
+    }
 }
 
-impl<'tcx> Visitor<'tcx> for ConstantExtractor {
+impl<'a, 'b, 'c, 'mir, 'tcx> Visitor<'tcx> for ConstantExtractor<'a, 'b, 'c, 'mir, 'tcx> {
     fn visit_constant(&mut self, constant: &mir::Constant<'tcx>) {
         self.super_constant(constant);
         match constant.literal {
             // already computed by rustc
             mir::Literal::Value { .. } => {}
-            mir::Literal::Item { .. } => {}, // FIXME: unimplemented
+            mir::Literal::Item { def_id, substs } => {
+                let item_ty = self.fncx.tcx.lookup_item_type(def_id).subst(self.fncx.tcx, substs);
+                if item_ty.ty.is_fn() {
+                    // unimplemented
+                } else {
+                    self.constant(def_id, substs, constant.span);
+                }
+            },
             mir::Literal::Promoted { index } => {
-                self.constants.push((ConstantId::Promoted { index: index }, constant.span));
+                if self.fncx.frame().promoted.contains_key(&index) {
+                    return;
+                }
+                let mir = self.mir.promoted[index].clone();
```

Comment:
This is cloning an entire `Mir` structure... we should try to find a way around that.